pkg/query/endpointset.go

```diff
@@ -302,39 +306,44 @@ func (e *EndpointSet) Update(ctx context.Context) {
 	e.updateMtx.Lock()
 	defer e.updateMtx.Unlock()
 
-	e.endpointsMtx.RLock()
-	endpoints := make(map[string]*endpointRef, len(e.endpoints))
-	for addr, er := range e.endpoints {
-		endpoints[addr] = er
-	}
-	e.endpointsMtx.RUnlock()
-
-	level.Debug(e.logger).Log("msg", "starting to update API endpoints", "cachedEndpoints", len(endpoints))
-
-	activeEndpoints := e.getActiveEndpoints(ctx, endpoints)
-	level.Debug(e.logger).Log("msg", "checked requested endpoints", "activeEndpoints", len(activeEndpoints), "cachedEndpoints", len(endpoints))
+	currentEndpoints := e.endpointRefs()
+	level.Debug(e.logger).Log("msg", "starting to update API endpoints", "cachedEndpoints", len(currentEndpoints))
 
-	stats := newEndpointAPIStats()
-
-	// Close endpoints which are not active this time (are not in active endpoints map).
-	for addr, er := range endpoints {
-		if _, ok := activeEndpoints[addr]; ok {
-			stats[er.ComponentType()][labelpb.PromLabelSetsToString(er.LabelSets())]++
+	endpointSpecs := make(map[string]*GRPCEndpointSpec)
+	for _, spec := range e.endpointSpec() {
+		if _, duplicate := endpointSpecs[spec.Addr()]; duplicate {
```

Comment:
Do we need this if we have https://github.com/thanos-io/thanos/blob/6d426d2b300ee192e617d7f5be1041aef308c798/cmd/thanos/query.go#L448 dedup ?

Reply:
Hm good question. I think EndpointSet should be self-contained and not rely on external assumptions. We might want to use it in other parts of the project (query frontend comes to mind since querier now has a gRPC API). So maybe it's best to stay defensive?

Reply:
I moved this into the EndpointSet constructor. We now make sure whatever is returned from specs() is deduplicated already.